prime-router/src/main/kotlin/cli/ProcessHl7Commands.kt

```diff
@@ -0,0 +1,236 @@
+                    segmentNumber
+                )
+
+                input is Composite && output is Composite -> {
+                    val inputComponents = input.components.filter { !it.isEmpty }
+                    val outputComponents = output.components.filter { !it.isEmpty }
+                    val inputExtraComponents = input.extraComponents
+                    val outputExtraComponents = output.extraComponents
+                    if (inputComponents.size != outputComponents.size) {
+                        echo("$segmentIndex-$segmentType number $segmentNumber did not match")
+                        return false
+                    } else if (inputExtraComponents.numComponents() != outputExtraComponents.numComponents()) {
+                        echo("$segmentIndex-$segmentType number $segmentNumber did not match")
+                        return false
+                    } else {
+                        inputComponents.zip(outputComponents).forEach { (i, o) ->
+                            if (!compareHl7Type(segmentIndex, i, o, segmentType, segmentNumber)) {
+                                echo("$segmentIndex-$segmentType number $segmentNumber did not match on $i, $o")
+                                return false
+                            }
+                        }
+                        return true
+                    }
+                }
+
+                else -> {
+                    return true
+                }
+            }
+        }
+
+        val mapNumOfSegment = mutableMapOf<String, Int>()
+
+        inputMap.forEach { (segmentIndex, segment) ->
+            // used to get the name of the segment as well as which number of that segment it is
+            val segmentIndexParts = segmentIndex.split("-")
+            val segmentType = segmentIndexParts[segmentIndexParts.size - 1]
+            val num = mapNumOfSegment[segmentType]
+            if (num == null) {
+                mapNumOfSegment[segmentType] = 1
+            } else {
+                mapNumOfSegment[segmentType] = num + 1
+            }
+
+            val outputSegment = outputMap[segmentIndex]
+            if (outputSegment == null) {
+                differences.add(Hl7Diff(segmentIndex, "Output missing segment"))
+                return@forEach
+            } else {
+                for (i in 1..segment.numFields()) {
+                    val inputFields = segment.getField(i)
+                    val outputFields = try {
+                        outputSegment.getField(i)
+                    } catch (ex: HL7Exception) {
+                        differences.add(
+                            Hl7Diff(
+                                segmentIndex,
+                                "Fields did not match, output was missing $segmentIndex.$i"
+                            )
+                        )
+                        continue
+                    }
+                    inputFields.forEachIndexed { index, input ->
```

Comment:
Similarly, could be `reduceIndexed`

Reply:
Or this. Not really familiar with reduce, in more complicated scenarios. Maybe we can pair on this?

Reply:
Yep let's pair on Monday